crates/oxc_cli/src/lint/runner_with_module_tree.rs

```diff
@@ -0,0 +1,324 @@
+use std::{
+    fs,
+    io::{BufWriter, Write},
+    path::{Path, PathBuf},
+    rc::Rc,
+    sync::{Arc, OnceLock},
+};
+
+use crossbeam_channel::{unbounded, Receiver, Sender};
+use miette::NamedSource;
+use oxc_allocator::Allocator;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+    Error, GraphicalReportHandler, Severity,
+};
+use oxc_linter::{FixResult, Fixer, Linter, RuleCategory, RuleEnum, RULES};
+use oxc_parser::{Parser, ParserReturn};
+use oxc_semantic::{SemanticBuilder, SemanticBuilderReturn};
+use oxc_span::{SourceType, VALID_EXTENSIONS};
+use rayon::prelude::*;
+use rustc_hash::FxHashSet;
+
+use super::{
+    resolver::{ResolveResult, Resolver, Resource},
+    AllowWarnDeny, LintOptions,
+};
+use crate::CliRunResult;
+
+#[derive(Clone)]
+struct LinterRuntimeData {
+    linter: Arc<Linter>,
+    visited: Arc<DashSet<PathBuf>>,
+    tx_error: Sender<(PathBuf, Vec<Error>)>,
+}
+
+pub struct LintRunnerWithModuleTree {
+    options: LintOptions,
+}
+use dashmap::DashSet;
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("File is too long to fit on the screen")]
+#[diagnostic(help("{0:?} seems like a minified file"))]
+pub struct MinifiedFileError(pub PathBuf);
+
+impl LintRunnerWithModuleTree {
+    pub fn new(options: LintOptions) -> Self {
+        Self { options }
+    }
+
+    pub fn print_rules() {
+        let mut stdout = BufWriter::new(std::io::stdout());
+        Linter::print_rules(&mut stdout);
+    }
+
+    fn derive_rules(options: &LintOptions) -> Vec<RuleEnum> {
+        let mut rules: FxHashSet<RuleEnum> = FxHashSet::default();
+
+        for (allow_warn_deny, name_or_category) in &options.rules {
+            let maybe_category = RuleCategory::from(name_or_category.as_str());
+            match allow_warn_deny {
+                AllowWarnDeny::Deny => {
+                    match maybe_category {
+                        Some(category) => rules.extend(
+                            RULES.iter().filter(|rule| rule.category() == category).cloned(),
+                        ),
+                        None => {
+                            if name_or_category == "all" {
```

Comment:
Is there an enum or something for this `"all"`? Or at least save it inside a constant to reduce the typos

Reply:
idk this part was there before I've touched it, look at commits before 228ad1a for more accurate diff

Reply:
This is my fault. Noted.